src/morale.cpp

```diff
@@ -526,38 +609,47 @@ void player_morale::update_stylish_bonus()
     int bonus = 0;
 
     if( stylish ) {
-        if( covered[bp_torso] ) {
-            bonus += 6;
-        }
-        if( covered[bp_head] ) {
-            bonus += 3;
-        }
-        if( covered[bp_eyes] ) {
-            bonus += 2;
-        }
-        if( covered[bp_mouth] ) {
-            bonus += 2;
-        }
-        if( covered[bp_leg_l] || covered[bp_leg_r] ) {
-            bonus += 2;
-        }
-        if( covered[bp_foot_l] || covered[bp_foot_r] ) {
-            bonus += 1;
+        const auto bp_bonus = [ this ]( body_part bp, int bonus ) -> int {
+            return (
+                body_parts[bp].covered_fancy > 0 ||
+                body_parts[opposite_body_part( bp )].covered_fancy > 0 ) ? bonus : 0;
+        };
+        bonus = std::min( super_fancy_bonus +
+                          bp_bonus( bp_torso,  6 ) +
+                          bp_bonus( bp_head,   3 ) +
+                          bp_bonus( bp_eyes,   2 ) +
+                          bp_bonus( bp_mouth,  2 ) +
+                          bp_bonus( bp_leg_l,  2 ) +
+                          bp_bonus( bp_foot_l, 1 ) +
+                          bp_bonus( bp_hand_l, 1 ), 20 );
+    }
+    set_permanent( MORALE_PERM_FANCY, bonus );
+}
+
+void player_morale::update_masochist_bonus()
+{
+    const bool amateur_masochist = has_mutation( "MASOCHIST" );
+    const bool advanced_masochist = has_mutation( "MASOCHIST_MED" ) || has_mutation( "CENOBITE" );
+    const bool any_masochist = amateur_masochist || advanced_masochist;
+
+    int bonus = 0;
+
+    if( any_masochist ) {
+        bonus = perceived_pain / 2.5;
+        if( amateur_masochist ) {
+            bonus = std::min( bonus, 25 );
         }
-        if( covered[bp_hand_l] || covered[bp_hand_r] ) {
-            bonus += 1;
+        if( took_prozac ) {
+            bonus = bonus / 3;
         }
-
-        bonus = std::min( bonus + super_fancy_bonus, 20 );
     }
-
-    add_permanent( MORALE_PERM_FANCY, bonus, bonus, true );
+    set_permanent( MORALE_PERM_MASOCHIST, bonus );
 }
 
 void player_morale::update_bodytemp_penalty()
```

Comment:
This function isn't really correct - consider someone in very hot boots, but with freezing hands. Your formula will sum it up to zero, while actually both hot and cold penalties should apply.